syntax = "proto2";
option cc_enable_arenas = true;

import "blickfeld/config/scan_pattern.proto";
import "blickfeld/data/scanline.proto";

package blickfeld.protocol.data;

/**
 * This section describes the contents of a point cloud frame.
 */
message Frame {
    optional uint64 id = 1;  // Incremental frame ID since startup of the device
    repeated Scanline scanlines = 2;  // Refer to [Scanline](#blickfeld.protocol.data.Scanline)
    optional uint64 start_time_ns = 3; // Unit: [s] â€“ start frame timestamp
    optional config.ScanPattern scan_pattern = 5;  // Refer to [ScanPattern](#blickfeld.protocol.config.ScanPattern)
    optional uint32 total_number_of_points = 6;  // Number of laser pulses emitted in this frame.

    /**
     * Number of returned points recorded:
     *
     * Each point of the [total_number_of_points](#blickfeld.protocol.data.Frame.total_number_of_points) can produce several returns.
     */
    optional uint32 total_number_of_returns = 7;
}